### PR TITLE
UseCases layout with associated components

### PR DESCRIPTION
**Pull request by wackerow**
## Description
Implements the UseCases layout, along with components used by each of these pages

- Implements `layouts/UseCases.tsx`, analogous to `templates/use-cases.tsx` from Gatsby repo
- Updated all image asset references for pages utilizing the `use-cases` layout
- Implemented `TableOfContents` component, which involved the addition of a markdown parser to generate the table of contents items.
- `generateTableOfContents` can be found in `src/components/TableOfContents/utils`, which takes a string of markdown content, parses out comments, and then filters down to only the header lines, and parses them into nested items to match the interface needed for the ToC component.

## Temporary changes that leave TODOs
- All i18n components have been disabled temporarily until i18n implementation ready
- `PageMetadata` component withheld until i18n implementation ready
- TableOfContents pointing to ids that are not yet being used in the actual header element (thus the links don't actually work)

## Preview URLs
- https://deploy-preview-33--ethereum-org-fork.netlify.app/defi
- https://deploy-preview-33--ethereum-org-fork.netlify.app/nft
- https://deploy-preview-33--ethereum-org-fork.netlify.app/dao
- https://deploy-preview-33--ethereum-org-fork.netlify.app/desci
- https://deploy-preview-33--ethereum-org-fork.netlify.app/refi
- https://deploy-preview-33--ethereum-org-fork.netlify.app/social-networks
- https://deploy-preview-33--ethereum-org-fork.netlify.app/decentralized-identity

### CODE REVIEW

File: public/content/refi/refi-regenerative-economy-diagram.png
Comment (pettinarip) [inline public/content/refi/refi-regenerative-economy-diagram.png:1]: Hmm what happened with this change?  I opened the file and it is empty 🤔

File: src/components/Glossary/GlossaryDefinition/index.tsx [hunk 1]
```diff
@@ -17,13 +17,16 @@ const GlossaryDefinition: React.FC<IProps> = ({ term, size = "md" }) => {
 
   const textStyles = size === "sm" ? { mb: 0 } : {}
 
+  // TODO: Re-enable with translations once intl implemented
   return (
     <Box>
-      <OldHeading as="h3" lineHeight={1.4} id={term} {...headingStyles}>
-        <Translation id={`${term}-term`} />
-      </OldHeading>
+      <Heading as="h3" lineHeight={1.4} id={term} {...headingStyles}>
```
Comment (pettinarip) [inline src/components/Glossary/GlossaryDefinition/index.tsx]: I appreciate these changes but I don't think we should include them in the scope of this PR and also in the migration since:
1. will make it more complex to keep in sync with the Gatsby repo since now we are adding new changes in the new repo that don't exist in the other repo
2. we might introduce styling bugs

Having said that, using the `Heading` component here will remove some default spacing that the `OldHeading` has.

File: src/components/Quiz/constants.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// Quiz Hub
+export const PROGRESS_BAR_GAP = "4px"
+export const PASSING_QUIZ_SCORE = 65
+export const USER_STATS_KEY = "quizzes-stats"
+export const INITIAL_QUIZ = "what-is-ethereum"
+export const TOTAL_QUIZ_QUESTIONS_ANSWERED = 100000
+export const TOTAL_QUIZ_AVERAGE_SCORE = 67.4
+export const TOTAL_QUIZ_RETRY_RATE = 15.6
```
Comment (pettinarip) [inline src/components/Quiz/constants.ts]: This was living in the `quizzes` page before, right?

File: src/lib/types.ts [hunk 3]
```diff
@@ -70,3 +70,33 @@ export type Lang =
   | "vi"
   | "zh"
   | "zh-tw"
+
+export type TranslationKey = string
+
+/**
+ * Quiz data types
```
Comment (pettinarip) [inline src/lib/types.ts:77]: nice to have:

IMO types in this files are more generic and global.

The quizzes types are very specific, I'd place them in a new `types` file inside the `Quiz` folder.

File: src/components/Quiz/utils.ts [hunk 4]
```diff
@@ -1,22 +1,21 @@
...
 import allQuizzesData, {
   ethereumBasicsQuizzes,
   usingEthereumQuizzes,
-} from "../../data/quizzes"
+} from "@/data/quizzes"
+import type { RawQuizzes } from "@/lib/interfaces"
 
 export const getTotalQuizzesPoints = () =>
-  Object.values(allQuizzesData)
+  Object.values(allQuizzesData as RawQuizzes)
```
Comment (pettinarip) [inline src/components/Quiz/utils.ts]: Not needed. `allQuizzesData` is already of type `RawQuizzes`

File: src/components/TableOfContents/TableOfContentsLink.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import React from "react"
-import { Link as GatsbyLink } from "gatsby"
+// import { Link as GatsbyLink } from "gatsby"
```
Comment (pettinarip) [inline src/components/TableOfContents/TableOfContentsLink.tsx]: lets remove this line as we are importing our custom `Link` component below.
```suggestion
```

File: src/components/TableOfContents/TableOfContentsMobile.tsx [hunk 6]
```diff
@@ -21,9 +20,6 @@ export interface IPropsTableOfContentsMobile {
 }
 
 const Mobile: React.FC<IPropsTableOfContentsMobile> = ({ items, maxDepth }) => {
-  // TODO: Replace with direct token implementation after UI migration is completed
-  const lgBp = useToken("breakpoints", "lg")
```
Comment (pettinarip) [inline src/components/TableOfContents/TableOfContentsMobile.tsx]: Related to the previous comment about doing these changes here. I wouldn't do them here.

File: src/components/TableOfContents/index.tsx [hunk 7]
```diff
@@ -41,9 +37,6 @@ const TableOfContents: React.FC<IProps> = ({
   isMobile = false,
   ...rest
 }) => {
-  // TODO: Replace with direct token implementation after UI migration is completed
-  const lgBp = useToken("breakpoints", "lg")
```
Comment (pettinarip) [inline src/components/TableOfContents/index.tsx]: out of scope

File: tsconfig.json
Comment (pettinarip) [inline tsconfig.json:1]: 🔥

File: src/components/TableOfContents/utils.ts [hunk 8]
```diff
@@ -56,3 +57,68 @@ export const outerListProps: ListProps = {
...
+ * @returns Array of `Item` objects parsed from the headings
+ */
+const addHeadingAsItem = (headings: Array<string>, h = 2): Array<Item> => {
+  const items: Array<Item> = []
+  const depths: number[] = headings.map(
+    (heading) => heading.match(/^#+/)?.[0].length ?? 0
+  )
+  depths.forEach((depth, i): void => {
+    if (depth > h) return
+    const headingItem: Item = parseHeadingToItem(headings[i])
```
Comment (pettinarip) [inline src/components/TableOfContents/utils.ts]: not needed
```suggestion
    const headingItem = parseHeadingToItem(headings[i])
```

File: src/components/TableOfContents/utils.ts [hunk 9]
```diff
@@ -56,3 +57,68 @@ export const outerListProps: ListProps = {
...
+  return items
+}
+
+/**
+ * Splits the content by lines and filters out lines that don't start with #s
+ * Calls `addHeadingAsItem` with array of Markdown headers to generate list of `Item` objects
+ * @param content Markdown content as a string (all lines)
+ * @returns List of `Item` objects parsed from the content, nested according to heading depth
+ */
+export const generateTableOfContents = (content: string): Array<Item> => {
```
Comment (pettinarip) [inline src/components/TableOfContents/utils.ts]: Nice!

File: src/components/UpgradeTableOfContents.tsx [hunk 10]
```diff
@@ -87,19 +73,12 @@ const ItemsList: React.FC<IPropsItemsList> = ({ items, depth, maxDepth }) => {
...
-}): JSX.Element
-function UpgradeTableOfContents(props: {
   maxDepth?: number
-  items: Array<Item>
-}) {
-  const { items, maxDepth = 1 } = props
-
+}
+const UpgradeTableOfContents: React.FC<IPropsToC> = ({ items, maxDepth = 1}) => {
+  console.log({yeah: items})
```
Comment (pettinarip) [inline src/components/UpgradeTableOfContents.tsx]: ✂️

File: src/layouts/UseCases.tsx [hunk 11]
```diff
@@ -1,3 +1,448 @@
...
+import { isLangRightToLeft } from "@/lib/utils/translations"
+import type { ChildOnlyProp, Lang } from "@/lib/types"
+
+const commonHeadingProps: HeadingProps = {
+  fontWeight: 700,
+  lineHeight: 1.4,
+}
+
+const Heading1 = (props: HeadingProps) => (
+  <Heading as="h1" {...commonHeadingProps} fontSize="2.5rem" {...props} />
```
Comment (pettinarip) [inline src/layouts/UseCases.tsx]: We should keep using the `OldHeading` or set a `mb` to keep the spacing consistent.
image

File: src/layouts/UseCases.tsx [hunk 12]
```diff
@@ -1,3 +1,448 @@
...
+      bottom={0}
+      zIndex={99}
+      p={8}
+      width="full"
+      {...props}
+    />
+  )
+}
+
+export const UseCasesLayout = ({ children, frontmatter, slug, tocItems }) => {
```
Comment (pettinarip) [inline src/layouts/UseCases.tsx]: Lets define the types for the layout

File: src/layouts/UseCases.tsx [hunk 13]
```diff
@@ -1,3 +1,448 @@
...
+  if (!frontmatter)
+    throw new Error("UseCasesLayout missing expected `frontmatter` prop")
+  if (!frontmatter?.title)
+    throw new Error("Required `title` property missing for UseCasesLayout")
```
Comment (pettinarip) [inline src/layouts/UseCases.tsx]: With the correct typing we shouldn't need these checks anymore

File: src/layouts/UseCases.tsx [hunk 14]
```diff
@@ -1,3 +1,448 @@
...
+        </ContentContainer>
+        <Hide above="lg">
+          <MobileButton>
+            <MobileButtonDropdown list={dropdownLinks} />
+          </MobileButton>
+        </Hide>
+      </Page>
+    </Box>
+  )
+}
```
Comment (pettinarip) [inline src/layouts/UseCases.tsx:460]: Are there any significant changes you remember doing in this file? 

Asking just to keep track of those changes and keep them in mind for future repo syncs.

**Review by pettinarip — CHANGES_REQUESTED** [review]
Great job overall @wackerow 

Good to see that the ToC is now working as well 💪🏼

File: src/components/TableOfContents/utils.ts
Comment (nhsz) [inline src/components/TableOfContents/utils.ts:1]: @wackerow utils should be defined inside `src/lib/utils`

**Review by nhsz — CHANGES_REQUESTED** [review]

**Comment by nhsz** [comment]
@wackerow noticing some visual differences (font weight, margins) vs our current design that apply to all added pages, eg: these are some buttons with missing margins on defi page (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)

Screen Shot 2023-10-04 at 12 11 48

Screen Shot 2023-10-04 at 12 11 58

Screen Shot 2023-10-04 at 12 12 16

Screen Shot 2023-10-04 at 12 12 27

---

These are some differences in font weight, margins

### PREVIEW (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)



### PROD (https://ethereum.org/en/defi/)



---

### PREVIEW (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)

1b

### PROD (https://ethereum.org/en/defi/)

2b

---

Maybe some could be related to recent changes pulled to theme @pettinarip?

File: src/components/TableOfContents/utils.ts
Comment (wackerow) [reply src/components/TableOfContents/utils.ts:1]: Okay... it was nested with the component in the Gatsby repo, are we trying to generally avoid doing that with this transition?

File: public/content/refi/refi-regenerative-economy-diagram.png
Comment (wackerow) [inline public/content/refi/refi-regenerative-economy-diagram.png:1]: File was not loading from `/public/use-cases/refi...png` so relocated next to markdown file (`/public/content/refi/refi...png`) which allowd it to load just fine

File: src/components/Glossary/GlossaryDefinition/index.tsx (hunk 1 above)
Comment (wackerow) [reply src/components/Glossary/GlossaryDefinition/index.tsx]: Yeah was noticing some of the layout bugs as a result, will revert

File: src/components/Quiz/constants.ts (hunk 2 above)
Comment (wackerow) [reply src/components/Quiz/constants.ts]: @pettinarip It was actually in `/src/constants.ts`... Probably should just move this back to constants now that I'm thinking about it. This isn't a pattern we're using yet of nesting `constant.ts` files other than in the Simulator (which we could also consider relocating to the root file separately)

File: src/components/Quiz/utils.ts
Comment (wackerow) [inline src/components/Quiz/utils.ts:1]: @nhsz Similar to your comment about `TableOfContents/utils.ts` belonging in `/src/lib/utils/`, do we also want to relocate this set of utilities in the same manner? I moved the ToC utils to `/src/lib/utils/toc.ts` and could similarly move these to `/src/lib/utils/quiz.ts`. cc: @pettinarip

File: src/components/UpgradeTableOfContents.tsx [hunk 15]
```diff
@@ -1,44 +1,13 @@
...
-  if (match) {
-    return match[2].toLowerCase()
-  }
-  console.warn("Missing custom ID on header: ", title)
-  return slugify(title)
-}
-
-const trimmedTitle = (title: string): string => {
-  const match = customIdRegEx.exec(title)
-  return match ? title.replace(match[1], "").trim() : title
```
Comment (wackerow) [inline src/components/UpgradeTableOfContents.tsx:22]: Repeated functions... these are now shared from `/src/lib/utils/toc.ts`

File: src/layouts/UseCases.tsx (hunk 11 above)
Comment (wackerow) [reply src/layouts/UseCases.tsx]: To prevent scope creep, will revert to use OldHeading (hah sorry, it's marked as deprecated so it gets crossed out in VSCode... hard to resist)

File: src/layouts/UseCases.tsx (hunk 12 above)
Comment (wackerow) [reply src/layouts/UseCases.tsx]: We have `frontmatter, slug, tocItems` defined locally inside `md.ts` right now as `type Items`... I'd suggest we rename to `interface PageContent` and pull that into `@/lib/interfaces` so we can reuse in places like right here... will go ahead with that, but let me know if you disagree

File: src/layouts/UseCases.tsx (hunk 14 above)
Comment (wackerow) [reply src/layouts/UseCases.tsx:460]: Main thing would probably be adjustments to the HeroContainer and the Image contained within when converting from GatsbyImage

File: src/lib/types.ts (hunk 3 above)
Comment (wackerow) [reply src/lib/types.ts:77]: Can't disagree... do we want to start bringing those types of adjustments into scope though? This is parity to Gatsby repo; I think I agree it's a nice to have if it happens, but wouldn't focus on this.

File: src/components/UpgradeTableOfContents.tsx (hunk 10 above)
Comment (wackerow) [reply src/components/UpgradeTableOfContents.tsx]: 😆 oops

File: src/components/Quiz/utils.ts [hunk 16]
```diff
@@ -1,19 +1,17 @@
-import { getLocaleForNumberFormat } from "../../utils/translations"
-import { Lang } from "../../utils/languages"
-
-import { CompletedQuizzes, QuizShareStats } from "../../types"
+import { getLocaleForNumberFormat } from "@/lib/utils/translations"
+import { CompletedQuizzes, type Lang, QuizShareStats } from "@/lib/types"
```
Comment (pettinarip) [inline src/components/Quiz/utils.ts:2]:
```suggestion
import type { CompletedQuizzes, Lang, QuizShareStats } from "@/lib/types"
```

File: src/components/TableOfContents/index.tsx [hunk 17]
```diff
@@ -13,18 +11,18 @@ import {
...
 
 import Mobile from "./TableOfContentsMobile"
 import ItemsList from "./ItemsList"
-import { getCustomId, type Item, outerListProps } from "./utils"
-import { trackCustomEvent } from "../../utils/matomo"
+import { getCustomId, outerListProps } from "@/lib/utils/toc"
+import type { ToCItem } from "@/lib/interfaces"
 
-export { Item }
+export { ToCItem }
```
Comment (pettinarip) [inline src/components/TableOfContents/index.tsx]: Wondering if this is even necessary now that we've moved it to `@/lib/interfaces`. I think we should consume it from there in all the other places.

File: src/lib/rehype/rehypeHeadingIds.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    const lastIndex = node.children.length - 1
+    const lastChild = node.children[lastIndex] as ElementContent
+    if ("value" in lastChild) {
+      const lastChildValue: string = lastChild.value as string
+      node.children[lastIndex] = {
+        type: "text",
+        value: trimmedTitle(lastChildValue),
+      }
+    }
+    // TODO: Add pseudo-element to heading to allow for anchor links
```
Comment (pettinarip) [inline src/lib/rehype/rehypeHeadingIds.ts:33]: Is there a card to track this? I'm not sure. Happy to create one if it is missing.

File: src/lib/rehype/rehypeHeadingIds.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import type { Plugin } from "unified"
+import type { ElementContent, Root } from "hast"
+import { getCustomId, trimmedTitle } from "@/lib/utils/toc"
+
+/**
+ * Parses DOM elements to find all headings, setting an `id` attribute on each one.
+ * The `getCustomId` Table of Contents utility functions is used to generate the `id`
+ * for each, allowing ToC links to match.
+ * The `trimmedTitle` function is used to remove any trailing `{#id}` from the heading
+ * @returns Plugin<[{}], Root>, a rehype plugin tht can be used in [...dynamic-md-pages].tsx
```
Comment (pettinarip) [inline src/lib/rehype/rehypeHeadingIds.ts]: 👍🏼

File: src/pages/[...slug].tsx [hunk 20]
```diff
@@ -72,7 +74,7 @@ export const getStaticPaths: GetStaticPaths = () => {
       return {
         params: {
           // Splitting nested paths to generate proper slug
-          slug: file.slug.split("/").slice(1),
+          slug: file.slug!.split("/").slice(1),
```
Comment (pettinarip) [inline src/pages/[...slug].tsx]: not needed anymore
```suggestion
          slug: file.slug.split("/").slice(1),
```

File: src/pages/[...slug].tsx [hunk 21]
```diff
@@ -86,6 +88,7 @@ export const getStaticProps: GetStaticProps<Props, Params> = async (
   const params = context.params!
   const markdown = getContentBySlug(params.slug.join("/"))
   const frontmatter = markdown.frontmatter as Frontmatter
+  const tocItems = markdown.tocItems as Array<ToCItem>
```
Comment (pettinarip) [inline src/pages/[...slug].tsx]: not needed anymore with the new `PageContent` type
```suggestion
  const frontmatter = markdown.frontmatter
  const tocItems = markdown.tocItems
```

File: src/lib/interfaces.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+  items?: Array<ToCItem>
+}
+
+/**
+ * Layout interface
+ */
+export interface PageContent {
+  slug: string
+  content: string
+  frontmatter: Frontmatter
```
Comment (pettinarip) [inline src/lib/interfaces.ts:62]: Thought that we were going to define the base required fields for `Frontmatter`, like `title` and `description`. Are we going to do that in a different PR?

File: src/lib/interfaces.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+export interface ToCItem {
+  title: string
+  url: string
+  items?: Array<ToCItem>
+}
+
+/**
+ * Layout interface
+ */
+export interface PageContent {
```
Comment (pettinarip) [inline src/lib/interfaces.ts:59]: Should we call this `MdPageContent`? since we will have react pages and those will not share the same props.

**Review by pettinarip — COMMENTED** [review]
Good job @wackerow I've added a few minor comments but overall it looks great.

These are the remaining tasks that I can think of that were raised from this PR (and we should tackle in separate PRs):
- Fix content images sizes in mobile
- Fix hero image size to match the prod version
- Add an anchor image in each heading (inside the new rehype plugin)

Am I missing something else?

File: src/lib/interfaces.ts (hunk 22 above)
Comment (pettinarip) [reply src/lib/interfaces.ts:62]: Oh just saw your changes in the other PR related to this. Ignore.

**Review by pettinarip — APPROVED** [review]

File: src/pages/[...slug].tsx (hunk 21 above)
Comment (wackerow) [reply src/pages/[...slug].tsx]: Agree... Working on a more robust front matter typing schema where we may need to add something back here, but can remove at the moment.

File: src/components/TableOfContents/index.tsx (hunk 17 above)
Comment (wackerow) [reply src/components/TableOfContents/index.tsx]: Agree, this isn't a pattern we use anywhere else... Just did a search and doesn't look like we are actually this export in any other file (they all import from `@/lib/interfaces`), so I'll just remove this line.

File: src/lib/interfaces.ts (hunk 23 above)
Comment (wackerow) [reply src/lib/interfaces.ts:59]: I'm down for that... I can address this separately when handing the front matter schema.

File: src/lib/rehype/rehypeHeadingIds.ts (hunk 18 above)
Comment (wackerow) [reply src/lib/rehype/rehypeHeadingIds.ts:33]: Not yet... I see you added this to your summary list though, so yeah we'll make cards for those.

**Comment by wackerow** [comment]
That list looks good @pettinarip , and agree we can tackle separately. Similarly, if we want to rename PageContent to MdPageContent that's fine with me, but I can just address that separately as well.

Should be ready for final set of eyes before merging. cc: @nhsz

File: src/lib/interfaces.ts (hunk 23 above)
Comment (pettinarip) [reply src/lib/interfaces.ts:59]: Sounds good!

File: src/lib/utils/toc.ts [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+  return emojiMatch ? trimmedTitle.replaceAll(emojiRegEx, "") : trimmedTitle
+}
+
+/**
+ * Removes {#...} from .md file so content can be parsed properly
+ * @deprecated Use `trimmedTitle` instead for each heading instead
+ * @param mdContent
+ * @returns string with {#id} removed
+ */
+export const removeAnchorLinks = (mdContent: string) =>
```
Comment (nhsz) [inline src/lib/utils/toc.ts]: @wackerow saw this util (`removeAnchorLinks`) was removed from `md.ts`, are we using it somewhere now? otherwise should be removed

**Review by nhsz — APPROVED** [review]
@wackerow great job! Left a comment regarding `removeAnchorLinks` util. Be sure to track related tasks yet to be done in Notion, mentioned in the comments

File: src/lib/utils/toc.ts (hunk 24 above)
Comment (wackerow) [reply src/lib/utils/toc.ts]: @nhsz We are not. This is primarily handled with `parseHeadingToItem` now. I just added a commit to remove this function... 

(Also updated `trimmedTitle` to be called `parseToCTitle` to be more explicit about what it's doing, and added comments for it.)

I think this is the last TODO for this PR.
